ITR/data/base_providers.py

```diff
@@ -614,16 +713,24 @@ def _get_trends(self, intensities: pd.DataFrame):
     def _extrapolate(self, trends: pd.DataFrame, projection_years: range, historic_data: pd.DataFrame) -> pd.DataFrame:
         projected_intensities = historic_data.loc[historic_data.index.intersection(trends.index)].copy()
         # We need to do a mini-extrapolation if we don't have complete historic data
+        # These columns are heterogeneous as to units, so don't try to use PintArrays
         for year in historic_data.columns.tolist()[:-1]:
-            m = projected_intensities[year + 1].apply(lambda x: np.isnan(x.m))
-            projected_intensities.loc[m, year + 1] = projected_intensities.loc[m, year] * (1 + trends.loc[m])
+            # FIXME: need a version of ITR.isnan that can see into Quantities
+            mask = ITR.isnan(projected_intensities[year + 1].map(lambda x: x.m if isinstance(x, pint.Quantity) else np.nan))
+            projected_intensities.loc[mask, year + 1] = projected_intensities.loc[mask, year] * (1 + trends.loc[mask])
 
         # Now the big extrapolation
         for year in projection_years:
             projected_intensities[year + 1] = projected_intensities[year] * (1 + trends)
+        # Clean up rows by converting NaN/None into Quantity(np.nan, unit_type)
+        # FIXME: Work-around for not-yet-filed Pandas issue whereby MultiIndex is converted into an Index of tuples
+        idx = projected_intensities.index
+        projected_intensities =  ITR.data.osc_units.asPintDataFrame(projected_intensities.T).T
```

Comment:
what's the point of double trnsponse?

Reply:
I just pushed changes that fix the FIXME and also make it more clear.  The ITR convention is to store company time-series data in rows, and the time-series should be compatible (but not necessarily homogeneous) with respect to units.  For example, a company may have reported CO2 emissions in tons for a series of years, and then mega-tons in later years.  Those units are compatible but not homogeneous.

Pandas has no concept of a datatype associated with rows, only columns.  This is explained in the _standardize function, but applies generally.  When we transpose the projected intensities, we create the opportunity to not only clean up and regularize the timeseries to being all the same units (due to PintArrays), but to also create a dataframe with proper pint[] dtype information on a per-column basis.  Alas, in this case that get's thrown away because we need the data back in row-major order, but we call this function to get homogeneous rows.  Hence the double-transpose (and no longer the save/restore of the index, because asPintDataFrame has been fixed).